substrate/primitives/runtime/src/lib.rs

```diff
@@ -1001,20 +1001,48 @@ macro_rules! assert_eq_error_rate_float {
 
 /// Simple blob to hold an extrinsic without committing to its format and ensure it is serialized
 /// correctly.
-#[derive(PartialEq, Eq, Clone, Default, Encode, Decode, DecodeWithMemTracking, TypeInfo)]
-pub struct OpaqueExtrinsic(Vec<u8>);
+#[derive(PartialEq, Eq, Clone, Default, Encode, Decode, DecodeWithMemTracking)]
+pub struct OpaqueExtrinsic(bytes::Bytes);
+
+impl TypeInfo for OpaqueExtrinsic {
+	type Identity = Self;
+	fn type_info() -> scale_info::Type {
+		scale_info::Type::builder()
+			.path(scale_info::Path::new("OpaqueExtrinsic", module_path!()))
+			.composite(
+				scale_info::build::Fields::unnamed()
+					.field(|f| f.ty::<Vec<u8>>().type_name("Vec<u8>")),
+			)
+	}
+}
 
 impl OpaqueExtrinsic {
+	/// Create a new instance of `OpaqueExtrinsic` from a `Vec<u8>`.
+	pub fn from_bytes(bytes: Vec<u8>) -> Self {
+		Self(bytes.into())
+	}
+
 	/// Convert an encoded extrinsic to an `OpaqueExtrinsic`.
-	pub fn from_bytes(mut bytes: &[u8]) -> Result<Self, codec::Error> {
+	pub fn try_from_encoded_extrinsic(mut bytes: &[u8]) -> Result<Self, codec::Error> {
 		Self::decode(&mut bytes)
 	}
+
+	/// Get the actual blob.
+	pub fn inner(&self) -> &[u8] {
+		&self.0
+	}
+}
+
+impl LazyExtrinsic for OpaqueExtrinsic {
+	fn decode_with_len(data: &[u8], _len: usize) -> Result<Self, codec::Error> {
+		Ok(Self(data.to_vec().into()))
```

Comment:
indeed it seems unnecessary to ensure `len == data.len()` because it is used to convert from lazy block to block. But I wonder if we should just check it anyway. Both way are good to me.